packages/cli/src/__tests__/execution-lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,526 @@
+			it('should send nodeExecuteAfter push event', async () => {
+				await hooks.executeHookFunctions('nodeExecuteAfter', [
+					nodeName,
+					taskData,
+					runExecutionData,
+				]);
+
+				expect(push.send).toHaveBeenCalledWith(
+					{ type: 'nodeExecuteAfter', data: { executionId, nodeName, data: taskData } },
+					pushRef,
+				);
+			});
+
+			it('should emit node-post-execute event', async () => {
+				await hooks.executeHookFunctions('nodeExecuteAfter', [
+					nodeName,
+					taskData,
+					runExecutionData,
+				]);
+
+				expect(eventService.emit).toHaveBeenCalledWith('node-post-execute', {
+					executionId,
+					workflow: workflowData,
+					nodeName,
+				});
+			});
+
+			it('should save execution progress when enabled', async () => {
+				workflowData.settings = { saveExecutionProgress: true };
+
+				await hooks.executeHookFunctions('nodeExecuteAfter', [
+					nodeName,
+					taskData,
+					runExecutionData,
+				]);
+
+				expect(executionRepository.findSingleExecution).toHaveBeenCalledWith(executionId, {
+					includeData: true,
+					unflattenData: true,
+				});
+			});
+
+			it('should not save execution progress when disabled', async () => {
+				workflowData.settings = { saveExecutionProgress: false };
+
+				await hooks.executeHookFunctions('nodeExecuteAfter', [
+					nodeName,
+					taskData,
+					runExecutionData,
+				]);
+
+				expect(executionRepository.findSingleExecution).not.toHaveBeenCalled();
+			});
+		});
+
+		describe('workflowExecuteBefore', () => {
+			it('should send executionStarted push event', async () => {
+				await hooks.executeHookFunctions('workflowExecuteBefore', [workflow, runExecutionData]);
+
+				expect(push.send).toHaveBeenCalledWith(
+					{
+						type: 'executionStarted',
+						data: {
+							executionId,
+							mode: executionMode,
+							retryOf,
+							workflowId: 'test-workflow-id',
+							workflowName: 'Test Workflow',
+							startedAt: now,
+							flattedRunData: '[{}]',
+						},
+					},
+					pushRef,
+				);
+			});
+
+			it('should not call eventService', async () => {
+				await hooks.executeHookFunctions('workflowExecuteBefore', [workflow, runExecutionData]);
+
+				expect(eventService.emit).not.toHaveBeenCalled();
+			});
+
+			it('should run workflow.preExecute external hook', async () => {
+				await hooks.executeHookFunctions('workflowExecuteBefore', [workflow, runExecutionData]);
+
+				expect(externalHooks.run).toHaveBeenCalledWith('workflow.preExecute', [
+					workflow,
+					executionMode,
+				]);
+			});
+		});
+
+		describe('workflowExecuteAfter', () => {
+			it('should send executionFinished push event', async () => {
+				await hooks.executeHookFunctions('workflowExecuteAfter', [successfulRun, {}]);
+				expect(eventService.emit).not.toHaveBeenCalled();
+				expect(push.send).toHaveBeenCalledWith(
+					{
+						type: 'executionFinished',
+						data: {
+							executionId,
+							rawData: '[{"resultData":"1"},{"runData":"2"},{}]',
```

Comment:
If we use `flatten` to generate this instead, we can mutate `successfulRun` at the top without breaking tests.